docs: add a `using viem` guide to Ignition (#4661)
NomicFoundation/hardhat: docs/src/content/ignition/docs/guides/viem.md

From the root directory of your Hardhat project run:

::::tabsgroup{options="npm 7+,npm 6,yarn"}

:::tab{value="npm 7+"}

```shell
npm install --save-dev @nomicfoundation/hardhat-ignition-viem
```

:::

:::tab{value="npm 6"}

```shell
npm install --save-dev @nomicfoundation/hardhat-ignition-viem
```

:::

:::tab{value=yarn}

```shell
yarn add --dev @nomicfoundation/hardhat-ignition-viem
```

:::

::::

Then [enable the plugin](../../../hardhat-runner/docs/guides/project-setup.md#plugins-and-dependencies) in your config file:

::::tabsgroup{options="TypeScript,JavaScript"}

:::tab{value="TypeScript"}

```typescript
import "@nomicfoundation/hardhat-ignition-viem";
```

:::

:::tab{value="JavaScript"}

```javascript
require("@nomicfoundation/hardhat-ignition-viem");
```

:::

::::

:::tip

Only **one** Hardhat Ignition package should be imported within your Hardhat config file. Viem and ethers support cannot both be enabled at the same time.

:::

## The Ignition object

The `@nomicfoundation/hardhat-plugin-viem` plugin adds an `ignition` object to the [Hardhat Runtime Environment](../../../hardhat-runner/docs/advanced/hardhat-runtime-environment.md).

> The `ignition` object exposes a `deploy` method, which takes an Ignition module, deploys it against the current network (in tests this will be the in-memory Hardhat network), and returns the results of the module as typed Viem contract instances.

Reviewer: ```suggestion
The `ignition` object exposes a `deploy` method, which takes an Ignition module, deploys it against the current network, like Hardhat Network, and returns the results of the module as typed Viem contract instances.
```

I corrected this as it wasn't always right.

If you are passing an artifact object to initialize a `Future` within a module, you will need to ensure the artifact's ABI is either declared with [const assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) or defined inline:

```typescript
const counterArtifact = {
  // ...
  abi: [...] as const, // <--- const assertion
}

const CounterModule = buildModule("Counter", (m) => {
  const counter = m.contract("Counter", CounterArtifact, [42]);

  return { counter };
});
```

If type inference isn't working for the returned contract, it is likely that the ABI type needs a `const` assertion or to be defined inline.

This is because the ABI type must be as narrow as possible for Viem to infer the correct type for the contract, and by default Typescript will broaden the types within object literals.

:::tip

Typescript doesn't support importing JSON as const, it will instead automatically broaden the type, breaking Viem's inference. The artifact must be defined inline within Typescript.

:::

See [Viem's type inference requirements](https://viem.sh/docs/typescript.html#type-inference) for more details.

> ## Sending transactions with a different account

Reviewer: Let's add a :::tip in the general testing guide, and create an issue to revisit the testing docs so that it is clearer that there are two "flavors"

> ## Sending transactions with a different account

Author: To clarify, sending transactions with a different account is the same in both the ethers and viem variants (the input parameters are the same across both versions, it is only the result that varies). I was suggesting moving the "Sending transactions with a different account" to the main testing section. Which on reflection, I am more convinced is the right move.

Can I check that I am following your comment correctly:

1. that we open an issue saying we should restructure the guides so that the current testing guide (assuming ethers) + viem stuck on as an extra, becomes, ethers testing guide + viem testing guide
2. But for the moment, add a tooltip to the general testing guide indicating there is a Viem option

> ## Sending transactions with a different account

Reviewer: > that we open an issue saying we should restructure the guides so that the current testing guide (assuming ethers) + viem stuck on as an extra, becomes, ethers testing guide + viem testing guide

Yes. Or maybe have two entirely different testing guides. We have to think about it.

> But for the moment, add a tooltip to the general testing guide indicating there is a Viem option

correct.

And given that situation, I don't think the duplication of the "sending transactions with a different account" is too bad.

> ## Sending transactions with a different account

Author: I have added the tooltip. I also duplicated the "Sending transactions with a different account" in the general test guide and made it work with ethers.

The `ignition.deploy` method will default to using the first account in Hardhat network's `accounts` array as the sender for all transactions.

You can change this by passing a `defaultSender` within the options object as a second argument to the `deploy` method:

```typescript
const [first, second] = await hre.viem.getWalletClients();

const result = await hre.ignition.deploy(ApolloModule, {
  defaultSender: second.account.address,
});
```

> ---

Reviewer: Let's add a title instead of `---` here.

Author: I switched it to `Hardhat Ignition and Hardhat Viem`.
